Release RC2 (#271)
JoshOrndorff/recipes: text/2-appetizers/3-errors.md

# Handling Errors

_[`pallets/adding-machine`](https://github.com/substrate-developer-hub/recipes/tree/master/pallets/adding-machine)_

> As we've mentioned before, in Substrate development, it is important to **Verify first, write

Reviewer: ```suggestion
As we've mentioned before, in Substrate development it is important to **Verify first, write
```

> last**. In this recipe, we'll create an adding machine checks for unlucky numbers (a silly example)

Reviewer: ```suggestion
last**. In this recipe, we'll create an adding machine that checks for unlucky numbers (a silly example)
```

}
```

Notice that the `Error` type always takes the generic parameter `T`. Notice also that we have
verified all preconditions, and thrown all possible errors before ever writing to storage.

## Throwing Errors with `.ok_or` and `.map_err`

In fact, the pattern of:

-   calling functions that returned a `Result` or `Option`, and

> -   checking if the result is `Some` or `Ok`. If not, returns from the function early with an error

Reviewer: ```suggestion
-   checking if the result is `Some` or `Ok` and if not, return from the function early with an error
```

}
```

Notice the pattern of `.ok_or(<Error<T>>::MyError)?;`. This is really handy when you have a function

> call that returns an `Option` and you expect there should be a value inside. If not, returns early

Reviewer: ```suggestion
call that returns an `Option` and you expect there should be a value inside. If not, return early
```
